doc/neps/nep-0037-array-module.rst

```diff
@@ -0,0 +1,332 @@
+
+For example, a library that supports arrays on both CPUs and GPUs might decide on which device to create a new arrays from functions like ``ones`` based on input arguments:
+
+.. code:: python
+
+    class Array:
+        def __array_module__(self, types, arrays):
+            useful_arrays = tuple(a in arrays if isinstance(a, Array))
+            if not useful_arrays:
+                return NotImplemented
+            prefer_gpu = any(a.prefer_gpu for a in useful_arrays)
+            return ArrayModule(prefer_gpu)
+
+    class ArrayModule:
+        def __init__(self, prefer_gpu):
+            self.prefer_gpu = prefer_gpu
+        
+        def __getattr__(self, name):
+            import base_module
+            base_func = getattr(base_module, name)
+            return functools.partial(base_func, prefer_gpu=self.prefer_gpu)
+
+This might be useful, but it's not clear if we really need it. Pint seems to get along OK without any explicit array creation routines (favoring multiplication by units, e.g., ``np.ones(5) * ureg.m``), and for the most part Dask is also OK with existing ``__array_function__`` style overides (e.g., favoring ``np.ones_like`` over ``np.ones``). Choosing whether to place an array on the CPU or GPU could be solved by `making array creation lazy <https://github.com/google/jax/pull/1668>`_.
+
+.. _appendix-design-choices:
+
+Appendix: design choices for API overrides
+------------------------------------------
+
+There is a large range of possible design choices for overriding NumPy's API. Here we discuss three major axes of the design decision that guided our design for ``__array_module__``.
+
+Opt-in vs. opt-out for users
+============================
+
+The ``__array_ufunc__`` and ``__array_function__`` protocols provide a mechanism for overriding NumPy functions *within NumPy's existing namespace*. This means that users need to explicitly opt-out if they do not want any overriden behavior, e.g., by casting arrays with ``np.asarray()``.
+
+In theory, this approach lowers the barrier for adopting these protocols in user code and libraries, because code that using the standard NumPy namespace is automatically compatible. But in practice, this hasn't worked out. For example, most well-maintained libraries that use NumPy follow the best practice of casting all inputs with ``np.asarray()``, which they would have to explicitly relax to use ``__array_function__``. Our experience has been that making a library compatible with a new duck array type typically requires at least a small amount of work to accomodate differences in the data model and operations that can be implemented efficiently.
```

Comment:
```suggestion
In theory, this approach lowers the barrier for adopting these protocols in user code and libraries, because code that uses the standard NumPy namespace is automatically compatible. But in practice, this hasn't worked out. For example, most well-maintained libraries that use NumPy follow the best practice of casting all inputs with ``np.asarray()``, which they would have to explicitly relax to use ``__array_function__``. Our experience has been that making a library compatible with a new duck array type typically requires at least a small amount of work to accomodate differences in the data model and operations that can be implemented efficiently.
```